exercises/custom-set/example.rs

```diff
@@ -1,149 +1,95 @@
-        let mut intersection: CustomSet<T> = CustomSet::new();
-        for element in self.elements.iter() {
-            if other.contains(element) {
-                intersection.insert(element.clone());
-            }
-        }
-        intersection
+    pub fn contains(&self, other: &T) -> bool {
+        self.collection.contains(other)
     }
 
-    pub fn union(&self, other: &CustomSet<T>) -> CustomSet<T> {
-        let mut union = CustomSet::new();
-        for element in self.elements.iter().chain(other.elements.iter()) {
-            union.insert(element.clone());
-        }
-        union
+    pub fn is_empty(&self) -> bool {
+        self.collection.is_empty()
     }
 
-    pub fn is_disjoint(&self, other: &CustomSet<T>) -> bool {
-        for element in self.elements.iter() {
-            if other.contains(element) {
-                return false;
-            }
-        }
-        true
+    pub fn size(&self) -> usize {
+        self.collection.len()
     }
 
-    pub fn is_subset(&self, other: &CustomSet<T>) -> bool {
-        for element in self.elements.iter() {
-            if !other.contains(element) {
-                return false;
-            }
-        }
-        true
-    }
-    
-    pub fn is_superset(&self, other: &CustomSet<T>) -> bool {
-        other.is_subset(self)
+    pub fn is_subset(&self, other: &Self) -> bool {
+        self.collection.iter().all(|x| other.contains(x))
     }
 
-    pub fn clear(&mut self) {
-        self.elements.clear();
+    pub fn is_disjoint(&self, other: &Self) -> bool {
+        !self.collection.iter().any(|x| other.contains(x))
     }
 
-    pub fn iter(&self) -> Iter<T>{
-        Iter { iter: self.elements.iter() }
+    pub fn intersection(&self, other: &Self) -> CustomSet<T> {
+        CustomSet::new(self.collection
+                           .iter()
+                           .filter(|c| other.contains(c))
+                           .map(|c| c.clone())
+                           .collect::<Vec<_>>())
     }
-}
 
-pub struct Iter<'a, T: 'a> {
-    iter: slice::Iter<'a, T>
-}
-
-impl<'a, T> IntoIterator for &'a CustomSet<T> where T: Eq + Clone {
-    type Item = &'a T;
-    type IntoIter = Iter<'a, T>;
-
-    fn into_iter(self) -> Iter<'a, T> {
-        self.iter()
+    pub fn union(&self, other: &Self) -> CustomSet<T> {
+        let mut s = CustomSet::new(vec![]);
+        for element in self.collection.iter() {
+            s.add(element.clone());
+        }
+        for element in other.collection.iter() {
+            s.add(element.clone());
+        }
+        s
     }
-}
-
-impl<'a, T> Iterator for Iter<'a, T> {
-    type Item = &'a T;
 
-    fn next(&mut self) -> Option<&'a T> {
-        self.iter.next()
+    pub fn difference(&self, other: &Self) -> CustomSet<T> {
+        CustomSet::new(self.collection
+                           .iter()
+                           .filter(|c| !other.contains(c))
+                           .map(|c| c.clone())
```

Comment:
Is this a time where https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.cloned would be useful?